subprojects/core/src/main/java/org/gradle/caching/internal/ShortCircuitingErrorHandlerBuildCacheServiceDecorator.java

```diff
@@ -64,29 +71,38 @@ public boolean load(BuildCacheKey key, BuildCacheEntryReader reader) {
     public void store(BuildCacheKey key, BuildCacheEntryWriter writer) {
         if (enabled.get()) {
             try {
+                LOGGER.debug("Storing entry {} in {} build cache", key, getRole());
                 super.store(key, writer);
             } catch (BuildCacheException e) {
+                LOGGER.warn(COULD_NOT_STORE_ENTRY_IN_BUILD_CACHE_LOG_MESSAGE, key, getRole(), e);
                 recordFailure();
                 // Assume its OK to not push anything.
+            } catch (Exception e) {
+                LOGGER.warn(COULD_NOT_STORE_ENTRY_IN_BUILD_CACHE_LOG_MESSAGE, key, getRole(), e);
+                disableBuildCache("a non-recoverable error was encountered.");
             }
         }
     }
 
     @Override
     public void close() throws IOException {
+        LOGGER.debug("Closing {} build cache", getRole());
         if (!enabled.get()) {
-            LOGGER.warn("The {} build cache was disabled during the build after encountering {} errors.",
-                getRole(), maxErrorCount
-            );
+            LOGGER.warn("The {} build cache was disabled during the build because " + disableMessage, getRole(), maxErrorCount);
         }
         super.close();
     }
 
     private void recordFailure() {
         if (remainingErrorCount.decrementAndGet() <= 0) {
-            if (enabled.compareAndSet(true, false)) {
-                LOGGER.warn("The {} build cache is now disabled because {} errors were encountered", getRole(), maxErrorCount);
-            }
+            disableBuildCache("{} recoverable errors were encountered.");
```

Comment:
I think we should just format this as we want it vs using SLF4J substitutions.  It makes the `LOGGER.warn` look wrong below (unless you know `message` has a `{}`).